### PR TITLE
add status reason to UnsatisfiedAssumption rejections

### PR DESCRIPTION
**Pull request by tybug**
closes https://github.com/HypothesisWorks/hypothesis/issues/3827.

Note that what we originally discussed in the issue has the knock-on effect of adding to the events of that test case, which shows up in e.g. [statistics mode](https://hypothesis.readthedocs.io/en/latest/details.html#test-statistics) reporting:

```
from hypothesis import *
from hypothesis.strategies import *

@given(integers())
def test_f(n):
    assume(n > 2)
```
```
sandbox.py::test_f:

  - during reuse phase (0.00 seconds):
    - Typical runtimes: < 1ms, of which < 1ms in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (0.04 seconds):
    - Typical runtimes: < 1ms, of which < 1ms in data generation
    - 99 passing examples, 0 failing examples, 96 invalid examples
    - Events:
      * 49.23%, invalid because: failed to satisfy assume()

  - Stopped because settings.max_examples=100
```

I'd call this desirable, though perhaps a tad confusing as it doesn't state which assume statement was unsatisfied. But I wonder if there was a good reason this hasn't been done yet.

cc @hgoldstein95

### CODE REVIEW

File: hypothesis-python/src/hypothesis/core.py
Comment (tybug) [inline hypothesis-python/src/hypothesis/core.py:999]: This explicitly mentions `assume()`, which isn't technically correct. `UnsatisfiedAssumption` might be raised by e.g. `reject()`. 

But given that `reject()` seems to be soft-deprecated in favor of `assume(False)`, and `assume()` is the vast majority of usages anyway, this felt a bit better than the more vague "failed to satisfy some assumption", which could be used to describe almost any usage of `mark_invalid`! I'm on the fence though.

Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/core.py:999]: I definitely don't want to leave potentially-misleading notes!  Here's an idea:

- Give `UnsatisfiedAssumption` a `reason` kwarg/attribute
- In `assume()` and `reject()`, pass the reason string.  
  - For bonus points, use [`inspect.stack(context=2)`](https://docs.python.org/3/library/inspect.html#inspect.stack) to work out where it was called from.
- If we don't have this information, don't add the event.

Comment (tybug) [reply hypothesis-python/src/hypothesis/core.py]: Something like this 9dbfbc7111062d6cdbef2d3d88f4c5e3c25233e3? I assume you meant `inspect.stack()[1]`, because `context` increases the amount of source code lines returned around that stackframe. We *could* use that to work out the exact call site of the `assume`, instead of just its calling function, but then multiline `assume` statements become a problem...

File: hypothesis-python/tests/cover/test_control.py
Comment (tybug) [inline hypothesis-python/tests/cover/test_control.py:38]: noticed in passing; I hope you don't mind. I couldn't find usages of `TD` (TestData, presumably?) elsewhere.

File: hypothesis-python/RELEASE.rst
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst:3]: Docs build is failing because we don't document `reject()`; it's not worth adding a deprecation warning but I'd prefer to steer people to `assume(False)` - fewer functions makes for a simpler and more-discoverable API.

File: hypothesis-python/tests/cover/test_statistical_events.py
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_statistical_events.py:274]: Linter is complaining that this loop variable (and below) is shadowing the import of `hypothesis.event`; rename?

File: hypothesis-python/src/hypothesis/errors.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/errors.py:27]: Apparently we have a test for whether we have a test for all error classes with custom init methods!

**Review by Zac-HD — APPROVED** [review]
CI is being nitpicky, but this looks great to me - merge once it's passing!

File: hypothesis-python/RELEASE.rst
Comment (tybug) [reply hypothesis-python/RELEASE.rst]: I didn't realize we don't document reject at all! I agree that we should be pushing `assume(False)` in general instead.

File: hypothesis-python/src/hypothesis/errors.py
Comment (tybug) [reply hypothesis-python/src/hypothesis/errors.py]: `test_exceptions_are_picklable` actually caught a bug due to this, which is that `UnsatisfiedAssumption` was no longer picklable after making `reason` a kwarg. To be honest, `reason` seems sensible as a posarg instead, so I turned it into one (which consequently fixed this issue).

I'm not particularly familiar with pickling issues like this, so maybe there is a more obvious solution.

**Comment by tybug** [comment]
Ouch. `inspect.stack` can get expensive in pathologic cases, as seen by `91.48s call     hypothesis-python/tests/cover/test_searchstrategy.py::test_example_raises_unsatisfiable_when_too_filtered`.

```python
from hypothesis import *
from hypothesis.strategies import *

@given(integers().filter(lambda x: False))
@settings(suppress_health_check=list(HealthCheck))
def f(n):
    pass

try:
    f()
except Exception:
    pass
```
image

**Comment by tybug** [comment]
Switched to `inspect.currentframe().f_back.f_code.co_name` which has a negligible impact.

**Comment by tybug** [comment]
I would hit the merge button here (you did tell me to!), but I changed some internals since, so I'm going to hold off. Free to yell at me if you think that's a bad call 🙂

**Comment by Zac-HD** [comment]
That's all about trusting your judgement, and if you want another review I'm happy to do so.  In this case it all looks good, so merging now!

**Pull request merged**
